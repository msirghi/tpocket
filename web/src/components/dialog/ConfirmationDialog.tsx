import React from 'react';
import Button from '@material-ui/core/Button';
import Dialog from '@material-ui/core/Dialog';
import DialogActions from '@material-ui/core/DialogActions';
import DialogContent from '@material-ui/core/DialogContent';
import DialogContentText from '@material-ui/core/DialogContentText';
import DialogTitle from '@material-ui/core/DialogTitle';
import LinearProgress from '@material-ui/core/LinearProgress';

interface IProps {
  open: boolean;
  toggleDialog: Function;
  onSubmit: () => void;
  isLoading: boolean;
  title: String;
}

export const ConfirmationDialog: React.FC<IProps> = (
  {
    open,
    toggleDialog,
    onSubmit,
    isLoading,
    title,
    children
  }) => {
  return (
    <div>
      <Dialog open={ open } onClose={ () => {
        toggleDialog(false);
      } } aria-labelledby="form-dialog-title">
        { isLoading && <LinearProgress/> }
        <DialogTitle id="form-dialog-title">
          { title }
        </DialogTitle>
        <DialogContent>
          <DialogContentText>
            { children }
          </DialogContentText>
        </DialogContent>
        <DialogActions>
          <Button
            onClick={ () => {
              toggleDialog(false)
            } }
            color="secondary"
          >
            Cancel
          </Button>
          <Button
            onClick={ onSubmit }
            color="primary"
          >
            Submit
          </Button>
        </DialogActions>
      </Dialog>
    </div>
  );
}
