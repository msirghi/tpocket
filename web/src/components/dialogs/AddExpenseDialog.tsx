import React, { useState } from 'react';
import Button from '@material-ui/core/Button';
import TextField from '@material-ui/core/TextField';
import Dialog from '@material-ui/core/Dialog';
import DialogActions from '@material-ui/core/DialogActions';
import DialogContent from '@material-ui/core/DialogContent';
import DialogContentText from '@material-ui/core/DialogContentText';
import DialogTitle from '@material-ui/core/DialogTitle';
import LinearProgress from '@material-ui/core/LinearProgress';
import { Category, useAddExpenseMutation } from '../../generated/graphql';
import Chip from '@material-ui/core/Chip';
import MenuItem from '@material-ui/core/MenuItem';
import Select from '@material-ui/core/Select';
import { startCase } from 'lodash';
import { useTheme } from '@material-ui/core/styles';
import { AlertMessage } from '../alerts/AlertMessage';
import { AlertType } from '../../commons/enums';

interface IProps {
  open: boolean;
  toggleDialog: Function;
  onSuccess: Function;
  categories: ({ __typename?: 'Category' | undefined } & Pick<Category, 'id' | 'name'>)[];
}

export const AddExpenseDialog: React.FC<IProps> = ({
  open,
  toggleDialog,
  categories,
  onSuccess
}) => {
  const theme = useTheme();
  const [addExpenseMutation] = useAddExpenseMutation();
  const [amount, setAmount] = useState<number | null>(null);
  const [error, setError] = useState<string | null>(null);
  const [isLoading, setLoading] = useState<boolean>(false);
  const [description, setDescription] = useState<string>('');
  const [selectedCategory, setSelectedCategory] = useState<number>(0);

  const onSubmit = async () => {
    if (amount && amount < 0) {
      setError('Amount cannot be negative.');
      return;
    }

    setError(null);
    setLoading(true);
    try {
      const response = await addExpenseMutation({
        variables: { amount: amount!, categoryId: selectedCategory }
      });
      onSuccess(response);
      toggleDialog(false);
      setLoading(false);
    } catch (e) {
      setError(e);
      setLoading(false);
    }
  };

  const handleSelectChange = (event: React.ChangeEvent<{ value: unknown }>) => {
    setSelectedCategory(event.target.value as number);
  };

  return (
    <div>
      <Dialog open={open} onClose={() => toggleDialog(false)} aria-labelledby='form-dialog-title'>
        {isLoading && <LinearProgress />}
        <DialogTitle id='form-dialog-title'>Add new expense</DialogTitle>
        <DialogContent>
          {error && (
            <div data-test='error-alert' style={{ marginBottom: '1rem' }}>
              <AlertMessage message={'Error'} type={AlertType.ERROR} />
            </div>
          )}
          <TextField
            data-testid='nameInput'
            data-test='amount-input'
            label='Amount'
            inputProps={{
              'data-testid': 'nameInput'
            }}
            type='number'
            onChange={(e) => setAmount(+e.target.value)}
            fullWidth
          />
          <TextField
            data-test='description-input'
            inputProps={{
              'data-testid': 'description-input'
            }}
            value={description}
            onChange={(e) => setDescription(e.target.value)}
            multiline
            rowsMax={4}
            margin='dense'
            label='Description'
            fullWidth
          />
          <div className={'mt-1'}>
            <DialogContentText>Category</DialogContentText>
            <div style={{ display: 'flex', justifyContent: 'space-around', flexWrap: 'wrap' }}>
              {categories.map(
                (category, idx) =>
                  idx < 5 && (
                    <Chip
                      style={{
                        background:
                          selectedCategory === category.id ? theme.palette.primary.main : '#f5f5f5',
                        color: selectedCategory === category.id ? '#fff' : '#000'
                      }}
                      onClick={() => setSelectedCategory(category.id)}
                      label={startCase(category.name)}
                      key={category.id}
                    />
                  )
              )}
            </div>

            {categories.length > 5 && (
              <Select
                labelId='demo-simple-select-label'
                id='demo-simple-select'
                inputProps={{
                  'data-testid': 'category-select'
                }}
                fullWidth
                data-testid={'category-select'}
                className={'mt-1'}
                value={selectedCategory || ''}
                onChange={handleSelectChange}
              >
                {categories.map(
                  (category, idx) =>
                    idx > 4 && (
                      <MenuItem data-test='menu-item' key={category.id} value={category.id}>
                        {startCase(category.name)}
                      </MenuItem>
                    )
                )}
              </Select>
            )}
          </div>
        </DialogContent>
        <DialogActions>
          <Button
            onClick={() => {
              toggleDialog(false);
            }}
            data-test='cancel-button'
            color='secondary'
          >
            Cancel
          </Button>
          <Button
            data-testid='submit'
            disabled={!amount || isLoading}
            onClick={onSubmit}
            color='primary'
            data-test='submit-button'
          >
            Submit
          </Button>
        </DialogActions>
      </Dialog>
    </div>
  );
};
