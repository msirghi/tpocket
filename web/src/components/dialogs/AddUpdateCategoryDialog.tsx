import React, { useEffect, useState } from 'react';
import Button from '@material-ui/core/Button';
import TextField from '@material-ui/core/TextField';
import Dialog from '@material-ui/core/Dialog';
import DialogActions from '@material-ui/core/DialogActions';
import DialogContent from '@material-ui/core/DialogContent';
import DialogContentText from '@material-ui/core/DialogContentText';
import DialogTitle from '@material-ui/core/DialogTitle';
import LinearProgress from '@material-ui/core/LinearProgress';
import { Category, useUpdateCategoryNameMutation } from '../../generated/graphql';
import { useSnackbar } from 'notistack';

interface IProps {
  open: boolean;
  toggleDialog: Function;
  onSubmit: Function;
  deselectCategory: Function;
  isLoading: boolean;
  selectedCategory: Category | null;
}

export const AddUpdateCategoryDialog: React.FC<IProps> = ({
  open,
  toggleDialog,
  onSubmit,
  isLoading,
  selectedCategory,
  deselectCategory
}) => {
  const [name, setName] = useState<string>('');
  const [description, setDescription] = useState<string>('');
  const [updateCategoryName] = useUpdateCategoryNameMutation();
  const { enqueueSnackbar } = useSnackbar();

  useEffect(() => {
    if (!open) {
      setName('');
      setDescription('');
    }
  }, [open]);

  useEffect(() => {
    if (selectedCategory) {
      setName(selectedCategory.name);
    }
  }, [selectedCategory]);

  const onUpdateCategoryName = async () => {
    try {
      const response = await updateCategoryName({ variables: { id: selectedCategory!.id, name } });
      if (response && response.data?.updateCategoryName) {
        enqueueSnackbar('Category updated', { variant: 'success' });
      }
    } catch (e) {
      enqueueSnackbar(e.graphQLErrors[0].message, { variant: 'error' });
    }
  };

  return (
    <div>
      <Dialog
        open={open}
        onClose={() => {
          deselectCategory();
          toggleDialog(false);
        }}
        aria-labelledby='form-dialog-title'
      >
        {isLoading && <LinearProgress />}
        <DialogTitle id='form-dialog-title'>
          {selectedCategory ? 'Edit category' : 'Add new category'}
        </DialogTitle>
        <DialogContent>
          <DialogContentText>
            {selectedCategory
              ? 'Update your category.'
              : 'Enter some information about new category.'}
          </DialogContentText>
          <TextField
            data-test='name-input'
            value={name}
            onChange={(e) => setName(e.target.value)}
            margin='dense'
            id='name'
            label='Name'
            type='email'
            fullWidth
          />
          <TextField
            data-test='description-input'
            value={description}
            onChange={(e) => setDescription(e.target.value)}
            multiline
            rowsMax={4}
            margin='dense'
            label='Description'
            fullWidth
          />
        </DialogContent>
        <DialogActions>
          <Button
            data-test='cancel-button'
            onClick={() => {
              deselectCategory();
              toggleDialog(false);
            }}
            color='secondary'
          >
            Cancel
          </Button>
          <Button
            data-test='submit-button'
            disabled={name.length === 0 || isLoading}
            onClick={async () => {
              if (selectedCategory) {
                toggleDialog();
                deselectCategory();
                await onUpdateCategoryName();
                return;
              }
              onSubmit({ name: name.trim(), description: description.trim() });
            }}
            color='primary'
          >
            {selectedCategory ? 'Update' : 'Create'}
          </Button>
        </DialogActions>
      </Dialog>
    </div>
  );
};
