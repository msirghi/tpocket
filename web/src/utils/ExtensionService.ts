const monthNames: Array<string> = [
  'January',
  'February',
  'March',
  'April',
  'May',
  'June',
  'July',
  'August',
  'September',
  'October',
  'November',
  'December'
];

const getMonthName = (monthNumber: number) => monthNames[monthNumber];

export default {
  getMonthName
};
