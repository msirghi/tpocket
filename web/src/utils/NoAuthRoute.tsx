import React from 'react';
import { Route, Redirect } from 'react-router-dom';
import { getAccessToken } from '../accessToken';

// @ts-ignore
export const NoAuthRoute = ({ Component, ...rest }) => (
  <Route
    {...rest}
    render={(props) => (getAccessToken() ? <Redirect to='/home' /> : <Component {...props} />)}
  />
);
