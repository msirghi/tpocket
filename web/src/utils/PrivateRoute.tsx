import React from 'react';
import { Route, Redirect } from 'react-router-dom';
import { getAccessToken } from '../accessToken';

// @ts-ignore
export const PrivateRoute = ({ Component, ...rest }) => (
  <Route
    {...rest}
    render={(props) => (getAccessToken() ? <Component {...props} /> : <Redirect to='/login' />)}
  />
);
