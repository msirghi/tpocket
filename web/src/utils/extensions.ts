const passwordStrength = require('check-password-strength');

const monthNames: Array<string> = [
  'January',
  'February',
  'March',
  'April',
  'May',
  'June',
  'July',
  'August',
  'September',
  'October',
  'November',
  'December',
];

export const getMonthName = (monthNumber: number) => monthNames[monthNumber];

export const validateEmailStr = (email: string) => {
  return !/^[A-Z0-9._%+-]+@[A-Z0-9.-]+\.[A-Z]{2,4}$/i.test(email);
};

export const isPasswordWeak = (password: string): boolean => {
  return passwordStrength(password).value === 'Weak';
};
