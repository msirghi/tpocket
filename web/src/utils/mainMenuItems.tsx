import React from "react";
import { IMenuItem } from "../interfaces/IMenuItem";
import HomeIcon from '@material-ui/icons/Home';
import AccountCircleIcon from '@material-ui/icons/AccountCircle';
import CategoryIcon from '@material-ui/icons/Category';
import HelpOutlineIcon from '@material-ui/icons/Help';
import ExitToAppIcon from '@material-ui/icons/ExitToApp';
import EmailIcon from '@material-ui/icons/Email';

export const mainMenuItems: Array<IMenuItem> = [
  {
    id: 1,
    link: '/home',
    name: 'Home',
    icon: <HomeIcon className={ 'light-grey-icon' }/>
  },
  {
    id: 2,
    link: '/account',
    name: 'Account',
    icon: <AccountCircleIcon className={ 'light-grey-icon ' }/>
  },
  {
    id: 3,
    link: '/categories',
    name: 'Categories',
    icon: <CategoryIcon className={ 'light-grey-icon' }/>
  }
];

export const secondaryMenuItems: Array<IMenuItem> = [
  {
    id: 1,
    link: '/help',
    name: 'Help',
    icon: <HelpOutlineIcon className={ 'light-grey-icon' }/>
  },
  {
    id: 2,
    link: '/writeus',
    name: 'Support',
    icon: <EmailIcon className={ 'light-grey-icon' }/>
  },
  {
    id: 3,
    link: '/logout',
    name: 'Sign out',
    icon: <ExitToAppIcon className={ 'light-grey-icon ' }/>
  }
];
