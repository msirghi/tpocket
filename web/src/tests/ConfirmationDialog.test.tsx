import React from 'react';
import { shallow, ShallowWrapper } from 'enzyme';
import { ConfirmationDialog } from '../components/dialogs/ConfirmationDialog';
import toJson from 'enzyme-to-json';
import { findFirstByTestAttr } from './utils/findByTestAttr';
import { Dialog } from '@material-ui/core';
import LinearProgress from '@material-ui/core/LinearProgress';

describe('ConfirmationDialog component', () => {
  let component: ShallowWrapper;
  let defaultProps = {
    open: true,
    toggleDialog: jest.fn(),
    onSubmit: jest.fn(),
    isLoading: false,
    title: 'Some title'
  };

  beforeEach(() => {
    component = shallow(<ConfirmationDialog {...defaultProps} />);
  });

  it('should render properly', () => {
    expect(toJson(component)).toMatchSnapshot();
  });

  it('should call toggleDialog on dialog close', () => {
    const cancelButton = findFirstByTestAttr(component, 'cancel-button');
    cancelButton.simulate('click');
    expect(defaultProps.toggleDialog).toBeCalledTimes(1);
  });

  it('should call onSubmit on submit button click', () => {
    const submitButton = findFirstByTestAttr(component, 'submit-button');
    submitButton.simulate('click');
    expect(defaultProps.onSubmit).toBeCalledTimes(1);
  });

  it('should call toggleDialog on dialog onClose event', () => {
    const dialog = component.find(Dialog);
    dialog.prop('onClose')();
    expect(defaultProps.toggleDialog).toBeCalledTimes(2);
  });

  it('should show LinearProgress if isLoading props is true', () => {
    const props = { ...defaultProps, isLoading: true };
    const wrapper = shallow(<ConfirmationDialog {...props} />);
    const linearProgress = wrapper.find(LinearProgress);
    expect(linearProgress).toHaveLength(1);
  });
});
