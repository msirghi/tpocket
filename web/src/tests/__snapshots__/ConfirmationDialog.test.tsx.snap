// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ConfirmationDialog component should render properly 1`] = `
<div>
  <WithStyles(ForwardRef(Dialog))
    aria-labelledby="form-dialog-title"
    onClose={[Function]}
    open={true}
  >
    <WithStyles(ForwardRef(DialogTitle))
      id="form-dialog-title"
    >
      Some title
    </WithStyles(ForwardRef(DialogTitle))>
    <WithStyles(ForwardRef(DialogContent))>
      <WithStyles(ForwardRef(DialogContentText)) />
    </WithStyles(ForwardRef(DialogContent))>
    <WithStyles(ForwardRef(DialogActions))>
      <WithStyles(ForwardRef(Button))
        color="secondary"
        data-test="cancel-button"
        onClick={[Function]}
      >
        Cancel
      </WithStyles(ForwardRef(Button))>
      <WithStyles(ForwardRef(Button))
        color="primary"
        data-test="submit-button"
        onClick={[MockFunction]}
      >
        Submit
      </WithStyles(ForwardRef(Button))>
    </WithStyles(ForwardRef(DialogActions))>
  </WithStyles(ForwardRef(Dialog))>
</div>
`;
